### PR TITLE
Add :exceptional-poll-delay-ms option, which causes a delay in dequeu…

### PR DESCRIPTION
**Pull request by stephencrampton**
…eing following an Exception.

### CODE REVIEW

File: src/com/climate/squeedo/sqs.clj [hunk 1]
```diff
@@ -302,24 +326,9 @@
...
-                  :wait-time-seconds poll-timeout-seconds
-                  :limit limit
-                  :attributes attributes
-                  :message-attribute-names message-attributes)
-         (map (fn [m]
-                (log/debugf "Dequeued from queue %s message %s" queue-name m)
-                (assoc m :queue-name queue-name))))
+    (apply dequeue* args)
     (catch Exception e
       (log/error e "Encountered exception dequeueing.")
```
Comment (stephencrampton) [inline src/com/climate/squeedo/sqs.clj]: This is highly dangerous.  If there is a network outage or SQS downtime, every thread on every host will emit logging messages.  We had a recent incident where one queue was inadvertently deleted in staging and 4 hosts generated 10,000 messages per second.  We risked overrunning our Splunk quota, not to mention all the noise in the logs.

Given there are typically multiple hosts with multiple threads using this library, I set the default delay to 10 seconds.  If it is a network hiccup, then one thread on one host will be delayed 10 seconds.  On the other hand, if there is a major outage, each thread on each host will issue a log every 10 seconds, which is still a lot of logging, but many orders of magnitude less than before this change.

File: test/com/climate/squeedo/sqs_consumer_test.clj [hunk 2]
```diff
@@ -76,14 +76,51 @@
...
+      (with-redefs [sqs/dequeue* (fn [& _]
+                                   (when @first-call
+                                     (reset! first-call false)
+                                     (throw (Exception. "thrown first time only")))
+                                   [(<!! test-chan)])]
+        (let [retry-ms 123
+              [message-chan _] (#'sqs-server/create-queue-listeners nil 1 2 1 retry-ms)
+              start-ms (System/currentTimeMillis)]
+          (<!! message-chan)
+          (is (> (- (System/currentTimeMillis) start-ms) retry-ms))
```
Comment (mtkp) [inline test/com/climate/squeedo/sqs_consumer_test.clj]: nice. probably wouldn't happen but maybe `>=`

File: test/com/climate/squeedo/sqs_consumer_test.clj [hunk 3]
```diff
@@ -96,6 +133,24 @@
           (wait-and-check 2)
           (is (false? (.full? buf)))
           (is (= 1 (.count buf)))
+          (close! message-chan)))))
+  (testing "Verify retry after specified ms"
+    (let [first-call (atom true)
```
Comment (mtkp) [inline test/com/climate/squeedo/sqs_consumer_test.clj]: another option for testing control flow is to just have a `calls` integer that you `(swap! calls inc)` when needed

File: test/com/climate/squeedo/sqs_test.clj [hunk 4]
```diff
@@ -53,6 +53,34 @@
...
+        (is (= expected
+               (sqs/dequeue {:client     {:dummy :client}
+                             :queue-name "queue-name"
+                             :queue-url  "URL"}
+                            :limit 2
+                            :attributes #{"Attribute"}
+                            :message-attributes #{"AttributeName"})))))
+    (with-redefs [sqs/receive (fn [& _]
+                                (throw (Exception. "SomeException")))]
+      (testing "dequeue swallows exceptions"
```
Comment (mtkp) [inline test/com/climate/squeedo/sqs_test.clj]: thanks for adding this test

File: test/com/climate/squeedo/sqs_consumer_test.clj [hunk 5]
```diff
@@ -96,6 +133,24 @@
...
+                                     (throw (Exception. "thrown first time only")))
+                                   (if @second-call
+                                     (do (reset! second-call false)
+                                         ["msg1" "msg2" "msg3" "msg4"])
+                                     []))]
+        (let [retry-ms 123
+              [message-chan _] (#'sqs-server/create-dedicated-queue-listeners nil 1 2 1 retry-ms)
+              start-ms (System/currentTimeMillis)]
+          (<!! message-chan)
+          (is (> (- (System/currentTimeMillis) start-ms) retry-ms))
```
Comment (mtkp) [inline test/com/climate/squeedo/sqs_consumer_test.clj]: same here... maybe use greater than or equal to

**Pull request merged**